File: agenkit/composition/conditional.py
"""Conditional agent composition pattern."""

from collections.abc import Callable
from dataclasses import dataclass

from agenkit.interfaces import Agent, Message

# Type alias for condition functions
Condition = Callable[[Message], bool]


@dataclass
class ConditionalRoute:
    """Represents a condition-agent pair."""

    condition: Condition
    agent: Agent


class ConditionalAgent(Agent):
    """Agent that routes messages to different agents based on conditions."""

    def __init__(self, name: str, default_agent: Agent):
        """Initialize conditional agent.

        Args:
            name: Name of this conditional agent
            default_agent: Agent to use when no condition matches
        """
        self._name = name
        self._routes: list[ConditionalRoute] = []
        self._default_agent = default_agent

    def add_route(self, condition: Condition, agent: Agent) -> None:
        """Add a conditional route.

        Args:
            condition: Function that returns True if this agent should be used
            agent: Agent to use when condition is met
        """
        self._routes.append(ConditionalRoute(condition=condition, agent=agent))

    @property
    def name(self) -> str:
        """Return the name of the conditional agent."""
        return self._name

    @property
    def capabilities(self) -> list[str]:
        """Return combined capabilities of all agents."""
        caps_set = set()

        # Add default agent capabilities
        if self._default_agent:
            caps_set.update(self._default_agent.capabilities)

        # Add route agent capabilities
        for route in self._routes:
            caps_set.update(route.agent.capabilities)

        caps = list(caps_set)
        caps.append("conditional")
        return caps

    async def process(self, message: Message) -> Message:
        """Route the message to the first agent whose condition is met.

        Args:
            message: Input message

        Returns:
            Response from the selected agent

        Raises:
            Exception: If no condition matches and no default agent configured
        """
        # Try each route in order
        for i, route in enumerate(self._routes):
            if route.condition(message):
                try:
                    result = await route.agent.process(message)

                    # Add metadata about routing decision
                    result.metadata["conditional_agent_used"] = route.agent.name
                    result.metadata["conditional_route"] = i + 1
                    return result

                except Exception as e:
                    raise Exception(f"Route {i+1} ({route.agent.name}) failed: {e}") from e

        # No condition matched, use default agent
        if not self._default_agent:
            raise Exception("No condition matched and no default agent configured")

        try:
            result = await self._default_agent.process(message)

            # Add metadata about using default
            result.metadata["conditional_agent_used"] = self._default_agent.name
            result.metadata["conditional_route"] = "default"
            return result

        except Exception as e:
            raise Exception(f"Default agent ({self._default_agent.name}) failed: {e}") from e

    def get_routes(self) -> list[ConditionalRoute]:
        """Return the conditional routes."""
        return self._routes

    def get_default_agent(self) -> Agent:
        """Return the default agent."""
        return self._default_agent


# Common condition helpers


def content_contains(substr: str) -> Condition:
    """Return a condition that checks if message content contains a substring.

    Args:
        substr: Substring to search for

    Returns:
        Condition function
    """

    def condition(message: Message) -> bool:
        return substr in message.content

    return condition


def role_equals(role: str) -> Condition:
    """Return a condition that checks if message role equals the given role.

    Args:
        role: Role to check for

    Returns:
        Condition function
    """

    def condition(message: Message) -> bool:
        return message.role == role

    return condition


def metadata_has_key(key: str) -> Condition:
    """Return a condition that checks if metadata contains a key.

    Args:
        key: Metadata key to check for

    Returns:
        Condition function
    """

    def condition(message: Message) -> bool:
        return key in message.metadata

    return condition


def metadata_equals(key: str, value) -> Condition:
    """Return a condition that checks if metadata key equals value.

    Args:
        key: Metadata key to check
        value: Expected value

    Returns:
        Condition function
    """

    def condition(message: Message) -> bool:
        return message.metadata.get(key) == value

    return condition


def and_conditions(*conditions: Condition) -> Condition:
    """Combine multiple conditions with AND logic.

    Args:
        *conditions: Conditions to combine

    Returns:
        Combined condition function
    """

    def condition(message: Message) -> bool:
        return all(cond(message) for cond in conditions)

    return condition


def or_conditions(*conditions: Condition) -> Condition:
    """Combine multiple conditions with OR logic.

    Args:
        *conditions: Conditions to combine

    Returns:
        Combined condition function
    """

    def condition(message: Message) -> bool:
        return any(cond(message) for cond in conditions)

    return condition


def not_condition(cond: Condition) -> Condition:
    """Negate a condition.

    Args:
        cond: Condition to negate

    Returns:
        Negated condition function
    """

    def condition(message: Message) -> bool:
        return not cond(message)

    return condition

File: agenkit-ts/src/composition/conditional.ts
/**
 * Conditional agent composition pattern.
 *
 * Routes messages to different agents based on conditions.
 */

import type { Agent, Message } from '../core/interfaces.js';

/**
 * Condition function type.
 *
 * Returns true if the message should be routed to the associated agent.
 */
export type Condition = (message: Message) => boolean;

/**
 * Represents a condition-agent pair.
 */
export interface ConditionalRoute {
  condition: Condition;
  agent: Agent;
}

/**
 * Agent that routes messages to different agents based on conditions.
 *
 * Evaluates conditions in order and routes to the first matching agent.
 * Falls back to default agent if no condition matches.
 *
 * @example
 * ```typescript
 * const conditional = new ConditionalAgent('router', defaultAgent);
 *
 * conditional.addRoute(
 *   (msg) => msg.content.includes('technical'),
 *   technicalAgent
 * );
 *
 * conditional.addRoute(
 *   (msg) => msg.content.includes('general'),
 *   generalAgent
 * );
 *
 * const result = await conditional.process(message);
 * ```
 */
export class ConditionalAgent implements Agent {
  readonly name: string;
  private routes: ConditionalRoute[] = [];
  private defaultAgent: Agent;

  /**
   * Create a new conditional agent.
   *
   * @param name - Name of this conditional agent
   * @param defaultAgent - Agent to use when no condition matches
   */
  constructor(name: string, defaultAgent: Agent) {
    this.name = name;
    this.defaultAgent = defaultAgent;
  }

  /**
   * Add a conditional route.
   *
   * @param condition - Function that returns true if this agent should be used
   * @param agent - Agent to use when condition is met
   */
  addRoute(condition: Condition, agent: Agent): void {
    this.routes.push({ condition, agent });
  }

  /**
   * Get combined capabilities of all agents.
   */
  get capabilities(): string[] {
    const capsSet = new Set<string>();

    // Add default agent capabilities
    if (this.defaultAgent.capabilities) {
      for (const cap of this.defaultAgent.capabilities) {
        capsSet.add(cap);
      }
    }

    // Add route agent capabilities
    for (const route of this.routes) {
      if (route.agent.capabilities) {
        for (const cap of route.agent.capabilities) {
          capsSet.add(cap);
        }
      }
    }

    const caps = Array.from(capsSet);
    caps.push('conditional');
    return caps;
  }

  /**
   * Route the message to the first agent whose condition is met.
   *
   * @param message - Input message
   * @returns Response from the selected agent
   * @throws Error if agent execution fails
   */
  async process(message: Message): Promise<Message> {
    // Try each route in order
    for (let i = 0; i < this.routes.length; i++) {
      const route = this.routes[i];

      if (route.condition(message)) {
        try {
          const result = await route.agent.process(message);

          // Add metadata about routing decision
          result.metadata = result.metadata || {};
          result.metadata.conditional_agent_used = route.agent.name;
          result.metadata.conditional_route = i + 1;

          return result;
        } catch (error) {
          const errorMessage =
            error instanceof Error ? error.message : String(error);
          throw new Error(
            `Route ${i + 1} (${route.agent.name}) failed: ${errorMessage}`,
          );
        }
      }
    }

    // No condition matched, use default agent
    try {
      const result = await this.defaultAgent.process(message);

      // Add metadata about using default
      result.metadata = result.metadata || {};
      result.metadata.conditional_agent_used = this.defaultAgent.name;
      result.metadata.conditional_route = 'default';

      return result;
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      throw new Error(
        `Default agent (${this.defaultAgent.name}) failed: ${errorMessage}`,
      );
    }
  }

  /**
   * Get the conditional routes.
   */
  getRoutes(): ConditionalRoute[] {
    return this.routes;
  }

  /**
   * Get the default agent.
   */
  getDefaultAgent(): Agent {
    return this.defaultAgent;
  }
}

// ==========================
// Common condition helpers
// ==========================

/**
 * Return a condition that checks if message content contains a substring.
 *
 * @param substr - Substring to search for
 * @returns Condition function
 */
export function contentContains(substr: string): Condition {
  return (message: Message): boolean => {
    return typeof message.content === 'string' && message.content.includes(substr);
  };
}

/**
 * Return a condition that checks if message role equals the given role.
 *
 * @param role - Role to check for
 * @returns Condition function
 */
export function roleEquals(role: string): Condition {
  return (message: Message): boolean => {
    return message.role === role;
  };
}

/**
 * Return a condition that checks if metadata contains a key.
 *
 * @param key - Metadata key to check for
 * @returns Condition function
 */
export function metadataHasKey(key: string): Condition {
  return (message: Message): boolean => {
    return message.metadata ? key in message.metadata : false;
  };
}

/**
 * Return a condition that checks if metadata key equals value.
 *
 * @param key - Metadata key to check
 * @param value - Expected value
 * @returns Condition function
 */
export function metadataEquals(key: string, value: unknown): Condition {
  return (message: Message): boolean => {
    return message.metadata ? message.metadata[key] === value : false;
  };
}

/**
 * Combine multiple conditions with AND logic.
 *
 * @param conditions - Conditions to combine
 * @returns Combined condition function
 */
export function andConditions(...conditions: Condition[]): Condition {
  return (message: Message): boolean => {
    return conditions.every((cond) => cond(message));
  };
}

/**
 * Combine multiple conditions with OR logic.
 *
 * @param conditions - Conditions to combine
 * @returns Combined condition function
 */
export function orConditions(...conditions: Condition[]): Condition {
  return (message: Message): boolean => {
    return conditions.some((cond) => cond(message));
  };
}

/**
 * Negate a condition.
 *
 * @param cond - Condition to negate
 * @returns Negated condition function
 */
export function notCondition(cond: Condition): Condition {
  return (message: Message): boolean => {
    return !cond(message);
  };
}
